pkg/skaffold/initializer/init.go

```diff
@@ -150,94 +172,136 @@ func DoInit(out io.Writer, c Config) error {
+	pairs := []BuilderImagePair{}
+	unresolvedImages := []string{}
+	for _, image := range images {
+		matchingConfigIndex := -1
+		for i, config := range buildConfigs {
+			if image != config.GetConfiguredImage() {
+				continue
+			}
+
+			// Found more than one match; can't auto-select.
+			if matchingConfigIndex != -1 {
+				matchingConfigIndex = -1
+				break
+			}
+			matchingConfigIndex = i
+		}
+
+		if matchingConfigIndex != -1 {
+			// Exactly one pair found; save the pair and remove from remaining build configs
+			pairs = append(pairs, BuilderImagePair{ImageName: image, Builder: buildConfigs[matchingConfigIndex]})
+			buildConfigs = append(buildConfigs[:matchingConfigIndex], buildConfigs[matchingConfigIndex+1:]...)
+		} else {
+			// No definite pair found, add to images list
+			unresolvedImages = append(unresolvedImages, image)
+		}
+	}
+	return pairs, unresolvedImages
+}
+
+func detectBuildFile(path string) ([]InitBuilder, error) {
+	// Check for Dockerfile
+	if docker.ValidateDockerfile(path) {
+		results := []InitBuilder{docker.Dockerfile(path)}
+		return results, nil
+	}
+
+	// TODO: Check for more builders
+
+	return nil, nil
+}
+
+func processCliArtifacts(artifacts []string) ([]BuilderImagePair, error) {
+	var pairs []BuilderImagePair
 	for _, artifact := range artifacts {
 		parts := strings.Split(artifact, "=")
 		if len(parts) != 2 {
 			return nil, fmt.Errorf("malformed artifact provided: %s", artifact)
 		}
-		pairs = append(pairs, dockerfilePair{
-			Dockerfile: parts[0],
-			ImageName:  parts[1],
+
+		pairs = append(pairs, BuilderImagePair{
+			Builder:   docker.Dockerfile(parts[0]),
+			ImageName: parts[1],
 		})
 	}
 	return pairs, nil
 }
 
-// For each image parsed from all k8s manifests, prompt the user for
-// the dockerfile that builds the referenced image
-func resolveDockerfileImages(dockerfiles []string, images []string) []dockerfilePair {
-	// if we only have 1 image and 1 dockerfile, don't bother prompting
-	if len(images) == 1 && len(dockerfiles) == 1 {
-		return []dockerfilePair{{
-			Dockerfile: dockerfiles[0],
-			ImageName:  images[0],
+// For each image parsed from all k8s manifests, prompt the user for the builder that builds the referenced image
+func resolveBuilderImages(buildConfigs []InitBuilder, images []string) []BuilderImagePair {
+	// If nothing to choose, don't bother prompting
+	if len(images) == 0 || len(buildConfigs) == 0 {
+		return []BuilderImagePair{}
+	}
+
+	// if we only have 1 image and 1 build config, don't bother prompting
+	if len(images) == 1 && len(buildConfigs) == 1 {
+		return []BuilderImagePair{{
+			Builder:   buildConfigs[0],
+			ImageName: images[0],
 		}}
 	}
-	pairs := []dockerfilePair{}
+
+	// Build map from choice string to builder config struct
+	choices := make([]string, len(buildConfigs))
+	choiceMap := make(map[string]InitBuilder, len(buildConfigs))
+	for i, buildConfig := range buildConfigs {
+		choice := buildConfig.Describe()
```

Comment:
Also worth mentioning on `Describe`'s godoc that the return result must be unique between artifacts